app/org/sagebionetworks/bridge/upload/IosSchemaValidationHandler2.java

```diff
@@ -0,0 +1,463 @@
+        this.healthDataDao = healthDataDao;
+    }
+
+    /** Upload Schema Service, used to get the schema corresponding to the upload. This is configured by Spring. */
+    @Autowired
+    public void setUploadSchemaService(UploadSchemaService uploadSchemaService) {
+        this.uploadSchemaService = uploadSchemaService;
+    }
+
+    /**
+     * Processes iOS data into health data records. iOS data comes from a third party, and we have no control over the
+     * data format. So our data validation needs to be as flexible as possible. Which means our error handling strategy
+     * is to write a validation message, and then attempt to recover. This may cause cascading errors further down the
+     * validation chain.
+     *
+     * @see org.sagebionetworks.bridge.upload.UploadValidationHandler#handle
+     */
+    @Override
+    public void handle(@Nonnull UploadValidationContext context)
+            throws UploadValidationException {
+        Map<String, JsonNode> jsonDataMap = context.getJsonDataMap();
+        Map<String, byte[]> unzippedDataMap = context.getUnzippedDataMap();
+        Upload upload = context.getUpload();
+        String uploadId = upload.getUploadId();
+        Study study = context.getStudy();
+        String studyId = study.getIdentifier();
+
+        // Add empty record builder and attachment map to the context. We'll fill these in as we need them.
+        HealthDataRecordBuilder recordBuilder = healthDataDao.getRecordBuilder();
+        context.setHealthDataRecordBuilder(recordBuilder);
+        Map<String, byte[]> attachmentMap = new HashMap<>();
+        context.setAttachmentsByFieldName(attachmentMap);
+
+        // health data records fields
+        recordBuilder.withHealthCode(upload.getHealthCode());
+        recordBuilder.withStudyId(studyId);
+        recordBuilder.withUploadDate(upload.getUploadDate());
+        recordBuilder.withUploadId(uploadId);
+
+        // create an empty object node in our record builder, which we'll fill in as we go
+        ObjectNode dataMap = BridgeObjectMapper.get().createObjectNode();
+        recordBuilder.withData(dataMap);
+
+        // Use info.json verbatim is the metadata.
+        JsonNode infoJson = getInfoJsonFile(context, uploadId, jsonDataMap);
+        recordBuilder.withMetadata(infoJson);
+
+        // validate and normalize filenames
+        validateInfoJsonFileList(context, uploadId, jsonDataMap, unzippedDataMap, infoJson, recordBuilder);
+        removeTimestampsFromFilenames(jsonDataMap);
+        removeTimestampsFromFilenames(unzippedDataMap);
+
+        // extract other data from info.json
+        JsonNode itemNode = infoJson.get(KEY_ITEM);
+        if (itemNode == null) {
+            // fall back to "identifier"
+            itemNode = infoJson.get(KEY_IDENTIFIER);
+        }
+        if (itemNode == null || itemNode.isNull()) {
+            throw new UploadValidationException("info.json is missing \"item\" field");
```

Comment:
This is done in an asynch worker. The upload validation task has code which catches these, writes these to the log, and updates the upload validation status accordingly:
https://github.com/Sage-Bionetworks/BridgePF/blob/develop/app/org/sagebionetworks/bridge/upload/UploadValidationTask.java#L70